prusti-viper/src/encoder/encoder.rs

```diff
@@ -365,112 +338,27 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
         has_spec_only_attr(self.env().tcx().get_attrs(def_id))
     }
 
-    /// Return the specification ids that are attached to `def_id` with one of the following
-    /// attributes:
-    /// * `prusti::pre_spec_id_ref="..."` for preconditions,
-    /// * `prusti::post_spec_id_ref="..."` for postconditions,
-    /// * `prusti::pledge_spec_id_ref="..."` for pledges.
-    fn get_procedure_spec_ids(&self, def_id: DefId) -> Vec<SpecIdRef> {
-        let mut spec_id_refs = vec![];
-        let attrs = self.env().tcx().get_attrs(def_id);
-
-        let parse_spec_id = |spec_id: String| -> SpecificationId {
-            spec_id.try_into().expect(
-                &format!("cannot parse the spec_id attached to {:?}", def_id)
-            )
-        };
-
-        spec_id_refs.extend(
-            read_prusti_attrs("pre_spec_id_ref", attrs).into_iter().map(
-                |raw_spec_id| SpecIdRef::Precondition(parse_spec_id(raw_spec_id))
-            )
-        );
-        spec_id_refs.extend(
-            read_prusti_attrs("post_spec_id_ref", attrs).into_iter().map(
-                |raw_spec_id| SpecIdRef::Postcondition(parse_spec_id(raw_spec_id))
-            )
-        );
-        spec_id_refs.extend(
-            read_prusti_attrs("pledge_spec_id_ref", attrs).into_iter().map(
-                |value| {
-                    let mut value = value.splitn(2, ":");
-                    let raw_lhs_spec_id = value.next().unwrap();
-                    let raw_rhs_spec_id = value.next().unwrap();
-                    let lhs_spec_id = if !raw_lhs_spec_id.is_empty() {
-                        Some(parse_spec_id(raw_lhs_spec_id.to_string()))
-                    } else {
-                        None
-                    };
-                    let rhs_spec_id = parse_spec_id(raw_rhs_spec_id.to_string());
-                    SpecIdRef::Pledge{ lhs: lhs_spec_id, rhs: rhs_spec_id }
-                }
-            )
-        );
-        debug!("Function {:?} has specification ids {:?}", def_id, spec_id_refs);
-        spec_id_refs
-    }
-
     /// Get the loop invariant attached to a function with a
     /// `prusti::loop_body_invariant_spec` attribute.
-    pub fn get_loop_specs(&self, def_id: DefId) -> Vec<SpecificationId> {
-        let attrs = self.env().tcx().get_attrs(def_id);
-        debug_assert!(has_prusti_attr(attrs, "loop_body_invariant_spec"));
-        read_prusti_attrs("spec_id", attrs).into_iter().map(
-            |raw_spec_id| raw_spec_id.try_into().expect(
-                &format!("cannot parse the spec_id attached to {:?}", def_id)
-            )
-        ).collect()
+    pub fn get_loop_specs(&self, def_id: DefId) -> Option<Vec<typed::Assertion<'tcx>>> {
```

Comment:
Would it be possible to return a reference here?

Reply:
It would be, but for the actual usages of both `get_loop_specs` and `get_procedure_specs`, the result is immediately cloned anyway (the procedure specs need to be put into a `typed::SpecificationSet::Procedure`, for example).

Reply:
I see.